File: JS_Submissions/binary_search/exercise2.js
/*
 Longest Subsequence With Limited Sum

Solution
You are given an integer array nums of length n, and an integer array queries of length m.

Return an array answer of length m where answer[i] is the maximum size of a subsequence that you can take from 
nums such that the sum of its elements is less than or equal to queries[i].

A subsequence is an array that can be derived from another array by deleting some or no elements without 
changing the order of the remaining elements.

 

Example 1:

Input: nums = [4,5,2,1], queries = [3,10,21]
Output: [2,3,4]
Explanation: We answer the queries as follows:
- The subsequence [2,1] has a sum less than or equal to 3. It can be proven that 2 is the maximum size of such a
subsequence, so answer[0] = 2.
- The subsequence [4,5,1] has a sum less than or equal to 10. It can be proven that 3 is the maximum size of 
such a subsequence, so answer[1] = 3.
- The subsequence [4,5,2,1] has a sum less than or equal to 21. It can be proven that 4 is the maximum size of 
such a subsequence, so answer[2] = 4.

Example 2:

Input: nums = [2,3,4,5], queries = [1]
Output: [0]
Explanation: The empty subsequence is the only subsequence that has a sum less than or equal to 1, so 
answer[0] = 0.
 

Constraints:

n == nums.length
m == queries.length
1 <= n, m <= 1000
1 <= nums[i], queries[i] <= 106
*/

/**
 * @param {number[]} nums
 * @param {number[]} queries
 * @return {number[]}
 */
var answerQueries = function(nums, queries) {
    nums.sort((a, b) => a - b);
    let ans = [];
    for (let i = 1; i < nums.length; i++) {
        nums[i] += nums[ i - 1];
    }
    
    const binarySearch = (arr, target) => {
        let left = 0, right = arr.length - 1;
        
        while(left <= right) {
            let mid = Math.floor((left + right) / 2);
            
            if(arr[mid] > target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }
    
    for (const query of queries) {
        let idx = binarySearch(nums, query);
        ans.push(idx);
    }
   
   
    return ans;
};

File: Rust_Submissions/sliding_window/exercise2.rs
/*
Max Consecutive Ones III

Solution
Given a binary array nums and an integer k, return the maximum number of consecutive 1's in the array if you can flip at most k 0's.

 

Example 1:

Input: nums = [1,1,1,0,0,0,1,1,1,1,0], k = 2
Output: 6
Explanation: [1,1,1,0,0,1,1,1,1,1,1]
Bolded numbers were flipped from 0 to 1. The longest subarray is underlined.
Example 2:

Input: nums = [0,0,1,1,0,0,1,1,1,0,1,1,0,0,0,1,1,1,1], k = 3
Output: 10
Explanation: [0,0,1,1,1,1,1,1,1,1,1,1,0,0,0,1,1,1,1]
Bolded numbers were flipped from 0 to 1. The longest subarray is underlined.
 

Constraints:

1 <= nums.length <= 105
nums[i] is either 0 or 1.
0 <= k <= nums.length
*/


pub fn longest_ones(nums: Vec<i32>, k: i32) -> i32 {
    let mut left = 0;
    let mut curr = 0;
    let mut ans = 0;
        
    for right in 0..nums.len() {
        if nums[right] == 0 {
            curr += 1;
        }
            
        while curr > k {
            if nums[left] == 0 {
                curr -= 1;
            }
                
                left += 1;
        }
            
        ans = ans.max((right - left + 1) as i32);
    }
    ans
}

// the work done in each loop iteration is amortized constant, so this algorithm has a runtime of O(n),
// where n is the length of nums, and O(1) space.